app/models/preserved_copy.rb

```diff
@@ -31,6 +31,13 @@ class PreservedCopy < ApplicationRecord
   validates :status, inclusion: { in: statuses.keys }
   validates :version, presence: true
 
+  scope :least_recent_version_audit, lambda { |last_checked_b4_date, storage_dir|
+    joins(:endpoint)
+      .where(endpoints: { storage_location: storage_dir })
+      .where('last_version_audit IS NULL or last_version_audit < ?', last_checked_b4_date)
+      .order('last_version_audit IS NOT NULL, last_version_audit ASC')
+  }
```

Comment:
please pardon my ignorance   (I didn't understand this before this refactor, either):

For performance or other reasons, do we need to break up the results returned into "chunks"?

Does this approach already "chunk" results?

If it's straight up Rails-foo, no need for tests of this ... just wondering.

Reply:
this doesn't chunk, but that's just because it just constructs the query w/o actually running it.  the query execution is handled by the caller's `.find_each` invocation on this query, which does do the chunking.

Reply:
ok, just did a bit of manual testing.  at first, i was worried that i was incorrect, because i ran some scopes in the rails console and got results immediately.  but then i noticed the same thing happened when i did a `join` without using a scope, and i know that returns an ActiveRecord relation and not an immediate result*.  so, i think what was happening was that pry was trying to immediately print the result of invoking `inspect` on the scope, and that was causing the query to be run.  for example, when i play around with some scopes i was experimenting with on friday, which are similar to the one in the PR, i got:
```ruby
# prevent pry from calling `inspect` on the return value of the scope, by having the result of the line be something other than the return value of the scope
[68] pry(main)> scope_query = PreservedCopy.by_endpoint_name('fixture_sr1').fixity_check_expired; p 'var is set'
"var is set"
=> "var is set"
...
[81] pry(main)> scope_query.class                                                                               
=> PreservedCopy::ActiveRecord_Relation
[82] pry(main)> 
```

so, i think scope should play just find with `find_each`, and should do the chunking thing correctly.  seems like it should be as composable as using `join` and `where` directly, since the scope returns an AR `Relation` object just like those do.

one slightly annoying thing i did notice when playing around with this stuff in the console:
```ruby
[5] pry(main)> PreservedCopy.by_endpoint_name('fixture_sr1').fixity_check_expired.find_each { |x| p x }
Scoped order is ignored, it's forced to be batch order.
  PreservedCopy Load (1.2ms)  SELECT  "preserved_copies".* FROM "preserved_copies" INNER JOIN "endpoints" ON "endpoints"."id" = "preserved_copies"."endpoint_id" INNER JOIN preserved_objects ON preserved_objects.id = preserved_object_id INNER JOIN preservation_policies ON preservation_policies.id = preserved_objects.preservation_policy_id AND extract(day from (CURRENT_TIMESTAMP - last_checksum_validation)*86400) > fixity_ttl OR last_checksum_validation IS NULL WHERE "endpoints"."endpoint_name" = $1 ORDER BY "preserved_copies"."id" ASC LIMIT $2  [["endpoint_name", "fixture_sr1"], ["LIMIT", 1000]]
#<PreservedCopy id: 1, version: 3, preserved_object_id: 1, endpoint_id: 1, created_at: "2017-12-13 20:19:49", updated_at: "2018-01-29 20:50:44", last_moab_validation: "2017-12-13 20:19:48", last_checksum_validation: nil, size: 2012695, status: "ok", last_version_audit: nil>
#<PreservedCopy id: 2, version: 3, preserved_object_id: 2, endpoint_id: 1, created_at: "2017-12-13 20:19:49", updated_at: "2017-12-13 20:19:49", last_moab_validation: "2017-12-13 20:19:49", last_checksum_validation: nil, size: 277588, status: "ok", last_version_audit: nil>
#<PreservedCopy id: 3, version: 2, preserved_object_id: 3, endpoint_id: 1, created_at: "2017-12-13 20:19:49", updated_at: "2017-12-13 20:19:49", last_moab_validation: "2017-12-13 20:19:49", last_checksum_validation: nil, size: 6570668, status: "ok", last_version_audit: nil>
=> nil
```

which indicates to me that we might want to move the ordering clause to the `find_each`, or have it in both places since some callers of the scope might want it, or something.  would be happy to ticket that consideration as follow on work.  doesn't seem essential to getting this PR merged.

* an ActiveRecord relation represents a query that can be run, but constructing it doesn't automatically run the query -- the query only gets run when the result set needs to be retrieved, e.g. when invoking something like `#find_each, `#first`, `#all`, or `#count` on the relation).

Reply:
the scopes referenced in the above comment can be seen here:
https://github.com/sul-dlss/preservation_catalog/compare/pres-copy-expired-fixity-scope?expand=1

they're pretty similar to the scope in this PR, i just happened to have that branch checked out already (because i was messing around for a bit with scopes we might want at the end of the day friday).

Reply:
cool - thanks for this super helpful detailed info.   I would say, yes please do create the ticket for adding ordering in for each loop.

Reply:
created https://github.com/sul-dlss/preservation_catalog/issues/537